File: dp/bookshop.py
n, x = map(int, input().split())
prices = list(map(int, input().split()))
values = list(map(int, input().split()))
ans = [[0] * (x + 1) for _ in range(n + 1)]

for i in range(1, n + 1):
  for j in range(x + 1):

    ans[i][j] = ans[i - 1][j]
    if j >= prices[i - 1]:
      ans[i][j] = max(ans[i][j], ans[i - 1][j - prices[i - 1]] + values[i - 1])

print(ans[n][x])

File: dp/bookshop.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
  int n, x;
  cin >> n >> x;
  vector<int> prices(n);
  for (int i = 0; i < n; i++)
  {
    cin >> prices[i];
  }
  vector<int> pages(n);
  for (int i = 0; i < n; i++)
  {
    cin >> pages[i];
  }

  vector<vector<int>> dp(n + 1, vector<int>(x + 1, 0));

  for (int i = 1; i <= n; i++)
  {
    for (int j = 0; j <= x; j++)
    {
      dp[i][j] = dp[i - 1][j];
      if (j - prices[i - 1] >= 0)
      {
        dp[i][j] = max(dp[i][j], dp[i - 1][j - prices[i - 1]] + pages[i - 1]);
      }
    }
  }

  cout << dp[n][x];
}